packages/react-server/src/ReactFizzServer.js

```diff
@@ -18,66 +23,819 @@ import {
   flushBuffered,
   close,
 } from './ReactServerStreamConfig';
-import {formatChunk} from './ReactServerFormatConfig';
-import {REACT_ELEMENT_TYPE} from 'shared/ReactSymbols';
+import {
+  writePlaceholder,
+  writeStartCompletedSuspenseBoundary,
+  writeStartPendingSuspenseBoundary,
+  writeStartClientRenderedSuspenseBoundary,
+  writeEndSuspenseBoundary,
+  writeStartSegment,
+  writeEndSegment,
+  writeClientRenderBoundaryInstruction,
+  writeCompletedBoundaryInstruction,
+  writeCompletedSegmentInstruction,
+  pushTextInstance,
+  pushStartInstance,
+  pushEndInstance,
+  createSuspenseBoundaryID,
+  createResponseState,
+} from './ReactServerFormatConfig';
+import {REACT_ELEMENT_TYPE, REACT_SUSPENSE_TYPE} from 'shared/ReactSymbols';
+import ReactSharedInternals from 'shared/ReactSharedInternals';
 
-type OpaqueRequest = {
-  destination: Destination,
-  children: ReactNodeList,
-  completedChunks: Array<Uint8Array>,
-  flowing: boolean,
-  ...
+import invariant from 'shared/invariant';
+
+const ReactCurrentDispatcher = ReactSharedInternals.ReactCurrentDispatcher;
+
+type SuspenseBoundary = {
+  +id: SuspenseBoundaryID,
+  rootSegmentID: number,
+  forceClientRender: boolean, // if it errors or infinitely suspends
+  parentFlushed: boolean,
+  pendingWork: number, // when it reaches zero we can show this boundary's content
+  completedSegments: Array<Segment>, // completed but not yet flushed segments.
+  byteSize: number, // used to determine whether to inline children boundaries.
+};
+
+type SuspendedWork = {
+  node: ReactNodeList,
+  ping: () => void,
+  blockedBoundary: Root | SuspenseBoundary,
+  blockedSegment: Segment, // the segment we'll write to
+  assignID: null | SuspenseBoundaryID, // id to assign to the content
```

Comment:
nit: Seems awkwardly named

Reply:
So the naming here is meant to indicate that this is a command. It's not like the ID of the parent and it's not meant to be the ID of the whole subtree if there are siblings, and it has special rules if a Suspense boundary gets in the way. It's just the first child, and if there is no appropriate first child it might insert a node. This is going to get messy regardless.